packages/react-utilities/src/compose/types.ts

```diff
@@ -48,3 +49,31 @@ export type SlotProps<TSlots extends BaseSlots, TProps, TRootProps extends React
 } & {
   root: TRootProps;
 };
+
+/**
+ * Helper type to convert the given props of type ShorthandProps into ObjectShorthandProps
+ */
+export type ResolvedShorthandProps<Props, ShorthandPropNames extends keyof Props> = Omit<Props, ShorthandPropNames> &
+  { [P in ShorthandPropNames]: Props[P] extends ShorthandProps<infer T> ? ObjectShorthandProps<T> : never };
+
+/**
+ * Helper type to mark the given props as required.
+ * Similar to Required<T> except it only requires a subset of the props.
+ */
+export type RequiredProps<Props, RequiredProps extends keyof Props> = Omit<Props, RequiredProps> &
+  { [P in RequiredProps]-?: Props[P] };
+
+/**
+ * Converts a components Props type to a State type:
+ * * Adds the 'ref' and 'as' types as required values
+ * * Ensures the specified ShorthandProps are of type ObjectShorthandProps<T>
+ * * Marks the given DefaultedProps as required (-?)
+ */
+export type ComponentState<
+  Props extends ComponentProps,
+  ShorthandProps extends keyof Props = never,
+  DefaultedProps extends keyof ResolvedShorthandProps<Props, ShorthandProps> = never
+> = RequiredProps<ResolvedShorthandProps<Props, ShorthandProps>, DefaultedProps> & {
+  as: React.ElementType;
+  ref: React.Ref<HTMLElement>;
```

Comment:
```suggestion
  ref: React.MutableRefObject<HTMLElement>;
```
```suggestion
  ref: React.RefObject<HTMLElement>;
```

It should be one of these types as `.Ref` also includes `.RefCallback` and `null`

Reply:
Are all components guaranteed/required to call `useMergedRefs` to get a RefObject?

Reply:
That would require every component to call `useMergedRefs` to create a ref object for its state. I don't think that's necessary for most components, is it? They can just forward the ref they're given. Or correct me if I'm misunderstanding...

Reply:
Yeah, that a good call 👍 However, as this type is shared and will be used for component where `useMergedRefs()` is used it will create types issues as `state.ref` according to them can be a function. For example:

```tsx
React.useEffect(() => {
  (state.ref as React.RefObject<HTMLElement>).current?.focus();
}, []);
```

[TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9uEAdqvG8jFgFxwDecSgV7Y8MAHTYCAHgASAFQCyAGQCiAGywgszGAD44AXzgBefoKzDMVCbkqcsUgBQBKI-Xqj84gK6osqgQEWPhOrqYGfPRwcE4cXOJCKOheElIA8gBGAFYhMHJKapraunou4rg+UJS6APziBBCVqK7RcPFYiZYVVTUSjc2tdIYANHAA2gC6LkA)